### PR TITLE
Matmul decide should_fold using has_out instead of grad_mode

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118617

Fixes https://github.com/pytorch/pytorch/issues/118548

### CODE REVIEW

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 1]
```diff
@@ -2014,7 +2017,7 @@ static Tensor _matmul_impl(
                    : tensor1.unsqueeze(0).mm(tensor2).squeeze_(0);
   } else if (dim_tensor1 == 2 && dim_tensor2 == 2) {
     return has_out ? at::mm_out(out, tensor1, tensor2) : tensor1.mm(tensor2);
-  } else if (should_fold(tensor1, tensor2)) {
+  } else if (should_fold(tensor1, tensor2, has_out)) {
```
Comment (lezcano) [inline aten/src/ATen/native/LinearAlgebra.cpp:2020]: Can we better write
```suggestion
  } else if (!has_out && should_fold(tensor1, tensor2)) {
```
?

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 2]
```diff
@@ -2014,7 +2017,7 @@ static Tensor _matmul_impl(
                    : tensor1.unsqueeze(0).mm(tensor2).squeeze_(0);
   } else if (dim_tensor1 == 2 && dim_tensor2 == 2) {
     return has_out ? at::mm_out(out, tensor1, tensor2) : tensor1.mm(tensor2);
-  } else if (should_fold(tensor1, tensor2)) {
+  } else if (!has_out && should_fold(tensor1, tensor2)) {
```
Comment (lezcano) [inline aten/src/ATen/native/LinearAlgebra.cpp:2020]: If we go with this fix, there is a bit of code in this branch that's now dead.

Comment (soulitzer) [reply aten/src/ATen/native/LinearAlgebra.cpp:2020]: Good catch. But now that I think about it, I actually prefer the previous version for a smaller blast radius lol

**Review by lezcano — APPROVED** [review]

**Comment by soulitzer** [comment]
It was kind of annoying to do, but just updated the decomp as well

**Pull request closed without merging**